runtime/compiler/optimizer/HandleRecompilationOps.cpp

```diff
@@ -0,0 +1,242 @@
+   //
+   if (!TR::Compiler->om.areValueTypesEnabled() || !comp()->isOSRAllowedForOperationsRequiringRecompilation())
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Handle Recompilation Operations is not enabled:  areValueTypesEnabled == %d; isOSRAllowedForOperationsRequiringRecompilation == %d\n", TR::Compiler->om.areValueTypesEnabled(), comp()->isOSRAllowedForOperationsRequiringRecompilation());
+         traceMsg(comp(), "Exiting HandleRecompilationOps\n");
+         }
+
+      return 0;
+      }
+
+   bool performedTransform = false;
+
+   for (TR::TreeTop *tt = comp()->getStartTree(); tt != NULL; tt = tt->getNextTreeTop())
+      {
+      TR::Node *node = tt->getNode();
+      performedTransform = performedTransform | visitNode(tt, node);
+      }
+
+   if (trace())
+     {
+     traceMsg(comp(), "Exiting HandleRecompilationOps\n");
+     }
+
+   return performedTransform ? 1 : 0;
+   }
+
+bool
+TR_HandleRecompilationOps::resolveCHKGuardsValueTypeOperation(TR::TreeTop *currTree, TR::Node *node)
+   {
+   TR::Node *resolveChild = node->getFirstChild();
+
+   if (resolveChild->getOpCodeValue() == TR::loadaddr)
+      {
+      // Check whether the ResolveCHK guards a loadaddr that is used in a newvalue
+      // operation.
+      //
+      TR::SymbolReference *loadAddrSymRef = resolveChild->getSymbolReference();
+
+      if (loadAddrSymRef->isUnresolved())
+         {
+         for (TR::TreeTopIterator iter(currTree->getNextTreeTop(), comp());
+              iter.currentNode()->getOpCodeValue() != TR::BBEnd;
+              ++iter)
+            {
+            TR::Node *nextNode = iter.currentNode();
+            if ((nextNode->getOpCode().isTreeTop() || nextNode->getOpCode().isCheck())
+                && nextNode->getNumChildren() > 0)
+               {
+               nextNode = nextNode->getFirstChild();
+               }
+
+            if (nextNode->getOpCodeValue() == TR::newvalue)
+               {
+               TR::Node *classAddr = nextNode->getFirstChild();
+
+               if (classAddr == resolveChild
+                   || (classAddr->getOpCodeValue() == TR::loadaddr
+                       && classAddr->getSymbolReference() == loadAddrSymRef))
+                  {
+                  return true;
+                  }
+               }
+            }
+         }
+      }
+   else if (resolveChild->getOpCode().isLoadVarOrStore() && resolveChild->getOpCode().isIndirect())
+      {
+      // Check whether the ResolveCHK guards a load or store operation on a field that is a Q type (value type)
+      //
+      TR::SymbolReference *symRef = resolveChild->getSymbolReference();
+
+      if (symRef->getCPIndex() != -1 && static_cast<TR_ResolvedJ9Method*>(_methodSymbol->getResolvedMethod())->isFieldQType(symRef->getCPIndex()))
+         {
+         return true;
+         }
+      }
+
+   return false;
+   }
+
+bool
+TR_HandleRecompilationOps::visitNode(TR::TreeTop *currTree, TR::Node *node)
```

Comment:
given that this is not recursive it might be cleaner just to pass the tree and pull out the node otherwise someone might think this might recurse...

Reply:
Quite right!  At first I was thinking it would be recursive, but then I never cleaned up the signature/name.  Fixed in commit bf856a82dd87a1eed2fc0dd8001d7df9b818b72d